sparse_ho/criterion.py

```diff
@@ -79,75 +69,76 @@ def compute_rmse(self, mask, dense, beta_star):
+
+        def get_v(mask, dense):
+            return 2 * (
+                X[np.ix_(self.idx_val, mask)].T @ (
+                    X[np.ix_(self.idx_val, mask)] @ dense - y[self.idx_val])) / len(self.idx_val)
+
         mask, dense, grad, quantity_to_warm_start = get_beta_jac_v(
-            self.model.X, self.model.y, log_alpha, self.model, self.get_v,
-            mask0=self.mask0, dense0=self.dense0,
+            X[self.idx_train, :], y[self.idx_train], log_alpha, model,
+            get_v, mask0=self.mask0, dense0=self.dense0,
             quantity_to_warm_start=self.quantity_to_warm_start,
             max_iter=max_iter, tol=tol, compute_jac=compute_jac, full_jac_v=True)
         self.mask0 = mask
         self.dense0 = dense
         self.quantity_to_warm_start = quantity_to_warm_start
-        mask, dense = self.model.get_primal(mask, dense)
-        val = self.value(mask, dense)
-        self.value_test(mask, dense)
+        mask, dense = model.get_primal(mask, dense)
+        val = self.get_mse_val(X, y, mask, dense)
+        # TODO put the following in a callback function
+        self.get_mse_test(mask, dense)
         self.compute_rmse(mask, dense, beta_star)
-        if self.convexify:
-            val += self.gamma_convex + np.sum(np.exp(log_alpha) ** 2)
-            grad += 2 * self.gamma_convex * np.exp(log_alpha)
+
         return val, grad
 
 
 class HeldOutLogistic():
     """Logistic loss on held out data
     """
 
-    def __init__(self, X_val, y_val, model, X_test=None, y_test=None):
+    def __init__(self, idx_train, idx_val, X_test=None, y_test=None):
         """
         Parameters
         ----------
         X_val : {ndarray, sparse matrix} of (n_samples, n_features)
             Validation data
         y_val : {ndarray, sparse matrix} of (n_samples)
             Validation target
-        model: object of the class Model (e.g. Lasso or Sparse logistic regression)
         X_test : {ndarray, sparse matrix} of (n_samples_test, n_features)
             Test data
         y_test : {ndarray, sparse matrix} of (n_samples_test)
             Test target
         """
-        self.X_val = X_val
-        self.y_val = y_val
+        self.idx_train = idx_train
+        self.idx_val = idx_val
         self.X_test = X_test
         self.y_test = y_test
-        self.model = model
 
         self.mask0 = None
         self.dense0 = None
         self.quantity_to_warm_start = None
         self.val_test = None
         self.rmse = None
 
-    def get_v(self, mask, dense):
-        temp = sigma(self.y_val * (self.X_val[:, mask] @ dense))
-        v = self.X_val[:, mask].T @ (self.y_val * (temp - 1))
-        v /= self.X_val.shape[0]
-        return v
-
-    def value(self, mask, dense):
-        val = np.sum(
-            np.log(1 + np.exp(-self.y_val * (self.X_val[:, mask] @ dense))))
-        val /= self.X_val.shape[0]
-        return val
+    @staticmethod
+    def value_outer_crit(X, y, mask, dense):
```

Comment:
not sure what it does, why it handles X and y being None, why criterion.get_value is not enough

Reply:
Because we still monitor the value on the test using this, and X_test can be None, this will be removed using a callback function